### ISSUE TITLE
How can I extract the commonsense features from a set of proposals for videos?

### ISSUE THREAD

**Issue by siyamsajeebkhan** [amanchadha/iPerceive#7]
Hi Aman, 
Congratulations for your excellent work. I am currently working on a DVC project. I read your paper and the idea of the CS is pretty interesting. I want to extract commonsense features from a set of proposals extracted from the video. I saw your code is based on images and you also suggested to make the dataset cocolike for applying iPerceive. Could you please provide me some help regarding this? 

Thank you!

**Comment by amanchadha** [comment]
Hi @siyamsajeebkhan,

Our codebase does tend to video input. All you have to do is feed individual frames of the video sequences as image inputs. For more details on how you can optimize commonsense feature generation for videos, check https://github.com/amanchadha/iPerceive#ive-heard-that-tasks-involving-video-need-serious-compute-horse-power-and-time-is-common-sense-generation-for-videos-computationally-expensive

For more on training iPerceive on a custom dataset, check https://github.com/amanchadha/iPerceive#how-do-i-train-iperceive-on-a-custom-dataset

Hope this helps!

**Comment by siyamsajeebkhan** [comment]
Thank you for your prompt reply. I still have one confusion. I saw you wrote the crux to using a custom dataset is to make it COCO like and I found out that to make a dataset COCO like I need to define the bounding boxes. But, you also wrote that iPerceive can also work without the bounding boxes. 

Now, let's say I have a set of proposals with start time and end time for each of the proposals. What I am thinking of is use ffmpeg to cut the relevant portions from the video and then feed each proposal frame by frame to the CS generator. Am I in the right direction? 

Also, in this case how can I leverage the power of iPerceive to extract CS features without providing the BB?

**Comment by amanchadha** [comment]
> Now, let's say I have a set of proposals with start time and end time for each of the proposals. What I am thinking of is use ffmpeg to cut the relevant portions from the video and then feed each proposal frame by frame to the CS generator. Am I in the right direction?

Yes, this appears to be the correct direction to explore.

> Also, in this case how can I leverage the power of iPerceive to extract CS features without providing the BB?

Check https://github.com/amanchadha/iPerceive#do-i-need-to-specify-bounding-boxes-for-my-dataset

**Comment by siyamsajeebkhan** [comment]
Thank you. 

> The following script is a push-button mechanism to use iPerceive for feature extraction with a pre-trained model (without any need for bounding-box data):
`run.sh`

Just to be sure, if I use this mechanism, then don't I need to make the customized dataset as per your instruction? 

> 1. Training on customized dataset

For learning iPerceive feature on your own dataset, the crux is to make your own dataset COCO-style (can refer to the data format in detection task) and design the dataloader file, for example coco.py and openimages.py. Here we provide an example for reference.

**Comment by amanchadha** [comment]
Correct! @siyamsajeebkhan

**Comment by siyamsajeebkhan** [comment]
Hi, 
I was able to run the `run.sh` file successfully and tested with a few images from the coco_2014_test dataset. I saw the output is saved as a `predictions.pth` file in a directory named 'results'. I was wondering how can I see the commonsense features? As far as I understood the commonsense features should have a size of N x 1024. I saw the outputs of the `run.sh` file are created as BoxList which contain the bounding boxes. But, how can I get the CS features from it?

**Comment by amanchadha** [comment]
You're on the right track. The `predictions.pth` file should be a PyTorch tensor dump of csfeatures. You can look at the code to see how it was generated (and then make your code to similarly load the pth and use it as csfeatures for downstream processing).

**Comment by siyamsajeebkhan** [comment]
Thank you! I guess, I have got the features now. Now my question is, how did you combine the features for the proposal videos? Did you concatenate all the CS features extracted from the frames of a given proposal video, hence forming a tensor of shape N X 1024?

**Comment by amanchadha** [comment]
Yes, concatenate those with the regular features. Further details are in the paper.

**Comment by siyamsajeebkhan** [comment]
Thank you @amanchadha for taking the time and clarifying my doubts. I wanted to ask you one more thing regarding the CS feature extraction from the videos. I am quoting some lines from your README:

> Yes, it is. However, to make the task of common-sense feature generation for videos tractable, we only generate common-sense features for a frame when we detect a change in the environmental "setting" going from one frame to the next in a particular localized event. Specifically, we check for changes in the set of object labels in a scene and only generate common-sense features if a change is detected; if not, we re-use the common-sense features from the last frame.

My question is, does your provided `run.sh` file automatically handle this thing or do I have to make some additional changes to make it compatible for the task of videos as described in the README. 

Thanks a lot for your cooperation. I really appreciate your help.

**Comment by amanchadha** [comment]
No worries @siyamsajeebkhan , glad to help! You will need to incorporate these changes to suit your setup. Note that these are optimizations and you do not need them to carry out essential task of CS features generation (which you've already done!).

**Comment by siyamsajeebkhan** [comment]
Hi @amanchadha , 
Thanks a lot and sorry for the delayed reply as I was in vacation. I would like to ask your opinion on designing the dataloader for working with the video files. As far as I have understood, I need to design my own dataloader for working with the videos to extract CS features. Now my question is, how did you design the dataloader for it? Did you first extract all the frames from the videos and used it as the dataset or did you load the video and pass it frame by frame? For working with the videos, is the annotation file necessary? Your suggestion would be really helpful for my task. 

Thanks a lot!
